impeller/renderer/backend/vulkan/render_pass_vk.cc

```diff
@@ -286,160 +287,124 @@ static bool UpdateBindingLayouts(const std::vector<Command>& commands,
-    write_set.setDstBinding(uniform.binding);
-    write_set.setDescriptorCount(1);
-    write_set.setDescriptorType(vk::DescriptorType::eUniformBuffer);
-    write_set.setPBufferInfo(&buffer_desc.back());
+      if (!encoder.Track(texture) ||
+          !encoder.Track(sampler.GetSharedSampler())) {
+        return false;
+      }
 
-    writes.push_back(write_set);
-  }
+      const SampledImageSlot& slot = bindings.sampled_images.at(index);
 
-  //----------------------------------------------------------------------------
-  /// Setup image descriptors.
-  ///
-  std::vector<vk::DescriptorImageInfo> image_descs;
-  for (const auto& [index, sampler_handle] : bindings.samplers) {
-    if (bindings.textures.find(index) == bindings.textures.end()) {
-      VALIDATION_LOG << "Missing texture for sampler: " << index;
-      return false;
-    }
+      vk::DescriptorImageInfo image_info;
+      image_info.imageLayout = vk::ImageLayout::eShaderReadOnlyOptimal;
+      image_info.sampler = sampler.GetSampler();
+      image_info.imageView = texture_vk.GetImageView();
 
-    auto texture = bindings.textures.at(index).resource;
-    const auto& texture_vk = TextureVK::Cast(*texture);
-    const SamplerVK& sampler = SamplerVK::Cast(*sampler_handle.resource);
+      vk::WriteDescriptorSet write_set;
+      write_set.dstSet = desc_set.value();
+      write_set.dstBinding = slot.binding;
+      write_set.descriptorCount = 1u;
+      write_set.descriptorType = vk::DescriptorType::eCombinedImageSampler;
+      write_set.pImageInfo = &(images[slot.binding] = image_info);
 
-    if (!encoder.Track(texture) || !encoder.Track(sampler.GetSharedSampler())) {
-      return false;
+      writes.push_back(write_set);
     }
 
-    const SampledImageSlot& slot = bindings.sampled_images.at(index);
+    return true;
+  };
 
-    vk::DescriptorImageInfo desc_image_info;
-    desc_image_info.setImageLayout(vk::ImageLayout::eShaderReadOnlyOptimal);
-    desc_image_info.setSampler(sampler.GetSamplerVK());
-    desc_image_info.setImageView(texture_vk.GetImageView());
-    image_descs.push_back(desc_image_info);
+  auto bind_buffers = [&allocator,  //
+                       &encoder,    //
+                       &buffers,    //
+                       &writes,     //
+                       &desc_set    //
+  ](const Bindings& bindings) -> bool {
+    for (const auto& [buffer_index, view] : bindings.buffers) {
+      const auto& buffer_view = view.resource.buffer;
+
+      auto device_buffer = buffer_view->GetDeviceBuffer(allocator);
+      if (!device_buffer) {
+        VALIDATION_LOG << "Failed to get device buffer for vertex binding";
+        return false;
+      }
 
-    vk::WriteDescriptorSet write_set;
-    write_set.setDstSet(desc_set);
-    write_set.setDstBinding(slot.binding);
-    write_set.setDescriptorCount(1);
-    write_set.setDescriptorType(vk::DescriptorType::eCombinedImageSampler);
-    write_set.setPImageInfo(&image_descs.back());
```

Comment:
For reference, this and the same for the buffer info above caused the dangling pointers. Inserting into the vector would invalidate the previous pointers.